plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/logging/KotlinPlaceholderCountMatchesArgumentCountInspection.kt

```diff
@@ -0,0 +1,208 @@
+                    argumentsCount = arrayArguments.size
+                    lastArgumentIsException =
+                        argumentsCount > 0 && arrayArguments.lastOrNull()?.getArgumentExpression()?.resolveType().isThrowable()
+                } else {
+                    return
+                }
+            }
+        }
+        val placeholderCount = countPlaceHolders(pattern, loggerType) ?: return
+        checkProblem(loggerType, lastArgumentIsException, argumentsCount, placeholderCount, holder, patternArgument)
+    })
+
+    private fun checkProblem(
+        loggerType: LoggerType,
+        lastArgumentIsException: Boolean,
+        argumentsCount: Int,
+        placeholderCount: Int,
+        holder: ProblemsHolder,
+        patternArgument: KtValueArgument
+    ) {
+        var actualArgumentsCount = argumentsCount
+        if (lastArgumentIsException) {
+            actualArgumentsCount--
+        }
+        when (loggerType) {
+
+            LoggerType.SLF4J_LOGGER -> {
+                if (placeholderCount == actualArgumentsCount) {
+                    return
+                }
+            }
+
+            LoggerType.LOG4J_BUILDER -> {
+                if ((placeholderCount == argumentsCount) ||
+                    (placeholderCount == argumentsCount - 1 && lastArgumentIsException)
+                ) {
+                    return
+                }
+            }
+
+            // if there is more than one argument and the last argument is an exception, but there is a placeholder for
+            // the exception, then the stack trace won't be logged.
+            LoggerType.LOG4J_LOGGER -> {
+                if ((placeholderCount == argumentsCount && (!lastArgumentIsException || argumentsCount > 1)) ||
+                    (placeholderCount == argumentsCount - 1 && lastArgumentIsException)
+                ) {
+                    return
+                }
+            }
+        }
+        registerProblem(placeholderCount, actualArgumentsCount, holder, patternArgument)
+    }
+
+    private fun registerProblem(
+        placeholderCount: Int,
+        argumentsCount: Int,
+        holder: ProblemsHolder,
+        patternArgument: KtValueArgument
+    ) {
+        if (placeholderCount < argumentsCount) {
+            holder.registerProblem(
+                patternArgument,
+                KotlinBundle.message("placeholder.count.matches.argument.count.more.problem.descriptor", argumentsCount, placeholderCount)
+            )
+        } else {
+            holder.registerProblem(
+                patternArgument,
+                KotlinBundle.message("placeholder.count.matches.argument.count.fewer.problem.descriptor", argumentsCount, placeholderCount)
+            )
+        }
+    }
+
+    private fun countPlaceHolders(pattern: KtExpression, loggerType: LoggerType): Int? {
+        val bindingContext = pattern.analyze(BodyResolveMode.PARTIAL)
+        val expectedType = pattern.getType(bindingContext) ?: return null
```

Comment:
Can `expectedType` be something different from `String`?

Reply:
Yes, changed to String type